src/adapter/src/coord/sequencer/inner/copy_from.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use mz_adapter_types::connection::ConnectionId;
+use mz_ore::cast::CastInto;
+use mz_persist_client::batch::ProtoBatch;
+use mz_repr::{CatalogItemId, Datum, RowArena};
+use mz_sql::plan::{self, CopyFromSource, HirScalarExpr};
+use mz_sql::session::metadata::SessionMetadata;
+use mz_storage_client::client::TableData;
+use mz_storage_types::oneshot_sources::OneshotIngestionRequest;
+use smallvec::SmallVec;
+use url::Url;
+
+use crate::coord::sequencer::inner::return_if_err;
+use crate::coord::{Coordinator, TargetCluster};
+use crate::optimize::dataflows::{prep_scalar_expr, EvalTime, ExprPrepStyle};
+use crate::session::{TransactionOps, WriteOp};
+use crate::{AdapterError, ExecuteContext, ExecuteResponse};
+
+impl Coordinator {
+    pub(crate) async fn sequence_copy_from(
+        &mut self,
+        ctx: ExecuteContext,
+        plan: plan::CopyFromPlan,
+        target_cluster: TargetCluster,
+    ) {
+        let plan::CopyFromPlan {
+            id,
+            source,
+            columns: _,
+            params: _,
+        } = plan;
+
+        let from_expr = match source {
+            CopyFromSource::Url(from_expr) => from_expr,
+            CopyFromSource::Stdin => {
+                unreachable!("COPY FROM STDIN should be handled elsewhere")
+            }
+        };
+
+        let eval_url = |from: HirScalarExpr| -> Result<Url, AdapterError> {
+            let style = ExprPrepStyle::OneShot {
+                logical_time: EvalTime::NotAvailable,
+                session: ctx.session(),
+                catalog_state: self.catalog().state(),
+            };
+            let mut from = from.lower_uncorrelated()?;
+            prep_scalar_expr(&mut from, style)?;
+
+            let temp_storage = RowArena::new();
+            let eval_result = from.eval(&[], &temp_storage)?;
+
+            if eval_result == Datum::Null {
+                coord_bail!("COPY FROM target value cannot be NULL");
+            }
+            let eval_string = match eval_result {
+                Datum::Null => coord_bail!("COPY FROM target value cannot be NULL"),
+                Datum::String(url_str) => url_str,
+                other => coord_bail!("programming error! COPY FROM target cannot be {other}"),
+            };
+
+            Url::parse(eval_string)
+                .map_err(|err| AdapterError::Unstructured(anyhow::anyhow!("{err}")))
```

Comment:
`coord_bail!` and `Unstructured` aren't the best because they give back unstructured errors, but they're so prevalent that it's probably fine.

Reply:
Yeah, I added a TODO to come back and clean this up